openwisp_firmware_upgrader/api/views.py

```diff
@@ -344,6 +349,67 @@ def get_object_or_none(self):
                 raise
 
 
+class UpgradeOperationCancelView(ProtectedAPIMixin, generics.GenericAPIView):
+    queryset = UpgradeOperation.objects.all()
+    serializer_class = serializers.Serializer
+    lookup_field = "pk"
+    organization_field = "device__organization"
+
+    CANCELLABLE_STATUS = UpgradeOperation._CANCELLABLE_STATUS
+    MAX_CANCELLABLE_PROGRESS = UpgradeOperation._MAX_CANCELLABLE_PROGRESS
+
+    def post(self, request, pk):
+        """Cancel an upgrade operation if conditions are met."""
+        try:
+            operation = self.get_object()
+        except Http404:
+            return self._error_response(
+                "Upgrade operation not found", status.HTTP_404_NOT_FOUND
+            )
+
+        try:
+            self._validate_cancellation(operation)
+            operation.cancel()
+            logger.info(
+                f"Upgrade operation {pk} cancelled successfully by user {request.user}"
+            )
+            return Response(
```

Comment:
let's move both logger and return outside of the try/except block, it could be written as `try/except/else` (so logger and return could be moved to the else block).